Add Tenant Page – List, View, Create, and Edit Tenants

oathar opened the pull request:
This PR adds a complete Tenant Management Page, accessible from the sidebar. It provides functionality to **list**, **view**, **edit**, and **create** tenants.

<img width="1901" height="954" alt="image" src="https://github.com/user-attachments/assets/9a4f82bc-9c84-4d26-80d9-5858781099d3" />


### Features Implemented

1.  **Tenant Listing:**  
- Fetches and displays a list of all tenants using the tenant API.
- Each tenant entry includes an action `👁️` to view detailed information.
> <img width="1187" height="801" alt="image" src="https://github.com/user-attachments/assets/b4bd3b9c-1e8c-4bfe-8bb4-708d4499398b" />
2.  **View Tenant Details:**  
  - Clicking the` 👁️ `icon opens a detailed view page for the selected tenant.
  - Displays all relevant tenant information in a user-friendly format.
> <img width="700" height="893" alt="image" src="https://github.com/user-attachments/assets/67d5d3d1-3565-4e04-8b1c-93b8dc808463" />
3. **Edit Tenant:**  
  - Users can edit existing tenant details via an edit modal.
  - All fields are editable except for the unique tenant `code`, which is immutable.
> <img width="725" height="826" alt="image" src="https://github.com/user-attachments/assets/21f0c41e-633f-4fb9-bbab-2260cec389a5" />
4. **Add New Tenant:**  
  - Users can create a new tenant by filling out a **dynamic form**.
  - On submission, the form data is validated and sent to the backend to create a new tenant.
> <img width="888" height="797" alt="image" src="https://github.com/user-attachments/assets/4d652e55-2b10-4685-8afa-e4fa21069981" />
### Technical Details
These features use the provided tenant APIs to` list all tenants`, `fetch details by code`, `create a new tenant`, `update existing ones`, and `add/remove domains to a tenant`. All API calls include proper loading states and error handling to ensure a smooth user experience. For tenant creation and editing, a **`dynamic form`** was implemented based on a `JSON schema`, an approach suggested by my **mentor**. @kiranharidas187  This makes the form flexible, easy to maintain, and scalable for future changes.

👉 Video link: https://drive.google.com/file/d/1WKtulaWXolw9miDO1Sp4xEQQVQlVMJFt/view?usp=sharing

oathar committed:
Add tenant page -list all tenants, view tenant details, create new tenant

oathar committed:
remove editTenant component

oathar committed:
Add Tenant Edit feature

oathar committed:
Add/remove  Tenant Domain

kiranharidas187 commented on src/api/tenantApi.js:
```diff
@@ -0,0 +1,116 @@
+import axios from 'axios'
+const BASE_URL = 'https://dev.elevate-apis.shikshalokam.org'
```
Base url should be in a .env file  file so that we can just modify it for every environemnts like dev, qa and prod. MAke sure the .env is not part of the PR by adding it to .gitignore. You can put a .env_sample file insted which will  have the key and value could be some sample  values rather than original value

kiranharidas187 commented on src/api/tenantApi.js:
```diff
@@ -0,0 +1,116 @@
+
+// Get all tenants
+export const getAllTenants = async () => {
+  const response = await axios.get(`${BASE_URL}/user/v1/tenant/list`, {
```
all api urls (eg /user/v1/tenant/list) should be in a single constant file so that its easy to modify at one place.

kiranharidas187 commented on src/api/tenantApi.js:
```diff
@@ -0,0 +1,116 @@
+// Get all tenants
+export const getAllTenants = async () => {
+  const response = await axios.get(`${BASE_URL}/user/v1/tenant/list`, {
+    headers: {
```
Instead of adding headers, axios, etc in every api request, figure out a generic way of doing this. For example, we need to add a new header tomorrow, and we need to add that header in all the places. Similarly, if we need to replace axios with another library, it should be just a one-place change.

kiranharidas187 commented on src/api/tenantApi.js:
```diff
@@ -0,0 +1,116 @@
+    })
+    return response.data
+  } catch (error) {
+    console.error('❌ Failed to fetch tenant details:', error?.response?.data || error.message)
```
Adding to the previous  comment, the error scenario alos should be handled in one place,

kiranharidas187 commented on src/components/ui/DynamicForm.jsx:
```diff
@@ -0,0 +1,175 @@
+        return (
+          <div key={name} className={commonClasses}>
+            <Label htmlFor={name}>{label}</Label>
+            {type === 'textarea' ? (
```
Instead of putting an if-else check here, can't we write separate definitions for text and textarea?? It makes the code more readable

kiranharidas187 commented on src/pages/Tenant/TenantCreateModal.jsx:
```diff
@@ -0,0 +1,291 @@
+    )
+  }
+
+  if (!formFields.length) {
```
What will happen if the API fails??

oathar committed:
Address mentor review comments:env config, API structure, error handling, code cleanup

oathar commented:
@kiranharidas187, Thank you for the feedback! I've made the changes based on your feedback: 

- Moved the API base URL to a `.env` file and added a `.env_sample`. 
- Put all API endpoints in one place for easy updates. 
- Created a single axios instance to handle API requests and errors. 
- Added proper error handling and user messages. 
- Cleaned up the DynamicForm code for better readability. 
- Removed console logs used for debugging. 

I tested tenant creation and editing, and everything works fine. Looking forward to your review.

oathar committed:
remove header duplication by setting headers in axios interceptor

oathar commented:
@kiranharidas187, I have now centralized all common headers (including Content-Type, x-auth-token, and x-tenant) in the axios interceptor, so there is no more header duplication in the tenant APIs. All manual header code has been removed from the API calls, and everything is working as expected.

Let me know if you have any further suggestions!
